### DISCUSSION TITLE
tensor.ToString() suggestion

### DISCUSSION THREAD

**Discussion by dayo05** [xamarin/TorchSharp#445] [Ideas]
In current, 
```cs
using static TorchSharp.torch;

var t = tensor(new float[] { 0, 1, 2, 3, 4, 5, 5 });
Console.WriteLine(t);
```
Prints `[7], type = Float32, device = cpu`

But In pytorch, printing tensor show its value.

Suggestion: make tensor.ToString() to show its value also.

**Comment by dayo05** [comment]
For example, 
```cs
Console.WriteLine(tensor(new float[] { 0, 1, 2, 3, 4, 5, 6 }));
```
prints data = [0, 1, 2, 3, 4, 5, 6], size = [7], type = Float32, device = cpu

**Comment by dayo05** [comment]
tensor.ToString(bool) supports that but, thats not look pretty

**Reply by Mattias-NCVIB** [reply]
I agree that something more would be postive, I use code that prints like this;
[[2,0,0],[0,1,0],[0,0,2]] and [[11,12,13],[14,15,16.55]]

which makes tests kinda simple;
new float[,] { { 11, 12, 13 }, { 14, 15, 16.55f } }.ToTensor2d().to_numpy_string().Should().Be("[[11,12,13],[14,15,16.55]]");

**Reply by NiklasGustafsson** [reply]
@dayo05: the ToString(bool,...) overload is meant to address this. Since many, many, tensors are gigantic, printing the values by default doesn't really work.

Please let me know what specifically it is that you think is is not pretty about the current formatting, and open an issue with the details, so that it's easier to track.

**Reply by mfagerlund** [reply]
For me, it's about writing tight tests and reviewing data to make sure it looks like I expect. I think the .str() method works great for some things but less well for others.

These examples are literally from tests of my helper methods;

            new float[,,] { { { 11, 12, 13 }, { 14, 15, 16.55f } }, { { 21, 22, 23 }, { 24, 25, 26.55f } } }.ToTensor3d().to_numpy_string().Should().Be("[[[11,12,13],[14,15,16.55]],[[21,22,23],[24,25,26.55]]]");
            new float[,] { { 11, 12, 13 }, { 14, 15, 16.55f } }.ToTensor2d().to_numpy_string().Should().Be("[[11,12,13],[14,15,16.55]]");
            new float[] { 14, 15, 16.55f }.ToTensor1d().to_numpy_string().Should().Be("[14,15,16.55]");
            new float[] { 5f }.ToTensor1d().to_numpy_string().Should().Be("[5]");

Using str() wouldn't work the same.

Now, for my needs, it's easy enough to provide my own extension method to print  the numpy strings.

**Reply by NiklasGustafsson** [reply]
I modeled the formatting after how the Julia REPL prints its tensors. That format is particularly helpful for large tensors, in particular those with more than 2 dimensions, where I find the NumPy style particularly challenging to interpret.

As you say, though, adding your own extension method is easy.

**Comment by GeorgeS2019** [comment]
[Related discussion](https://github.com/dotnet/TorchSharp/pull/680#issuecomment-1205657199)
